File: Sources/PDFjet/CoreFont.swift
/**
 * CoreFont.swift
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
import Foundation

public class CoreFont {
    public static let COURIER = 1
    public static let COURIER_BOLD = 2
    public static let COURIER_OBLIQUE = 3
    public static let COURIER_BOLD_OBLIQUE = 4
    public static let HELVETICA = 5
    public static let HELVETICA_BOLD = 6
    public static let HELVETICA_OBLIQUE = 7
    public static let HELVETICA_BOLD_OBLIQUE = 8
    public static let TIMES_ROMAN = 9
    public static let TIMES_BOLD = 10
    public static let TIMES_ITALIC = 11
    public static let TIMES_BOLD_ITALIC = 12
    public static let SYMBOL = 13
    public static let ZAPF_DINGBATS = 14

    var name: String?
    var bBoxLLx: Int16?
    var bBoxLLy: Int16?
    var bBoxURx: Int16?
    var bBoxURy: Int16?
    var underlinePosition: Int16?
    var underlineThickness: Int16?
    var metrics: [[Int16]]?

    public init(_ coreFont: Int) {
        if coreFont == CoreFont.COURIER {
            self.name = Courier.name
            self.bBoxLLx = Courier.bBoxLLx
            self.bBoxLLy = Courier.bBoxLLy
            self.bBoxURx = Courier.bBoxURx
            self.bBoxURy = Courier.bBoxURy
            self.underlinePosition = Courier.underlinePosition
            self.underlineThickness = Courier.underlineThickness
            self.metrics = Courier.metrics
        } else if coreFont == CoreFont.COURIER_BOLD {
            self.name = Courier_Bold.name
            self.bBoxLLx = Courier_Bold.bBoxLLx
            self.bBoxLLy = Courier_Bold.bBoxLLy
            self.bBoxURx = Courier_Bold.bBoxURx
            self.bBoxURy = Courier_Bold.bBoxURy
            self.underlinePosition = Courier_Bold.underlinePosition
            self.underlineThickness = Courier_Bold.underlineThickness
            self.metrics = Courier_Bold.metrics
        } else if coreFont == CoreFont.COURIER_OBLIQUE {
            self.name = Courier_Oblique.name
            self.bBoxLLx = Courier_Oblique.bBoxLLx
            self.bBoxLLy = Courier_Oblique.bBoxLLy
            self.bBoxURx = Courier_Oblique.bBoxURx
            self.bBoxURy = Courier_Oblique.bBoxURy
            self.underlinePosition = Courier_Oblique.underlinePosition
            self.underlineThickness = Courier_Oblique.underlineThickness
            self.metrics = Courier_Oblique.metrics
        } else if coreFont == CoreFont.COURIER_BOLD_OBLIQUE {
            self.name = Courier_BoldOblique.name
            self.bBoxLLx = Courier_BoldOblique.bBoxLLx
            self.bBoxLLy = Courier_BoldOblique.bBoxLLy
            self.bBoxURx = Courier_BoldOblique.bBoxURx
            self.bBoxURy = Courier_BoldOblique.bBoxURy
            self.underlinePosition = Courier_BoldOblique.underlinePosition
            self.underlineThickness = Courier_BoldOblique.underlineThickness
            self.metrics = Courier_BoldOblique.metrics
        } else if coreFont == CoreFont.HELVETICA {
            self.name = Helvetica.name
            self.bBoxLLx = Helvetica.bBoxLLx
            self.bBoxLLy = Helvetica.bBoxLLy
            self.bBoxURx = Helvetica.bBoxURx
            self.bBoxURy = Helvetica.bBoxURy
            self.underlinePosition = Helvetica.underlinePosition
            self.underlineThickness = Helvetica.underlineThickness
            self.metrics = Helvetica.metrics
        } else if coreFont == CoreFont.HELVETICA_BOLD {
            self.name = Helvetica_Bold.name
            self.bBoxLLx = Helvetica_Bold.bBoxLLx
            self.bBoxLLy = Helvetica_Bold.bBoxLLy
            self.bBoxURx = Helvetica_Bold.bBoxURx
            self.bBoxURy = Helvetica_Bold.bBoxURy
            self.underlinePosition = Helvetica_Bold.underlinePosition
            self.underlineThickness = Helvetica_Bold.underlineThickness
            self.metrics = Helvetica_Bold.metrics
        } else if coreFont == CoreFont.HELVETICA_OBLIQUE {
            self.name = Helvetica_Oblique.name
            self.bBoxLLx = Helvetica_Oblique.bBoxLLx
            self.bBoxLLy = Helvetica_Oblique.bBoxLLy
            self.bBoxURx = Helvetica_Oblique.bBoxURx
            self.bBoxURy = Helvetica_Oblique.bBoxURy
            self.underlinePosition = Helvetica_Oblique.underlinePosition
            self.underlineThickness = Helvetica_Oblique.underlineThickness
            self.metrics = Helvetica_Oblique.metrics
        } else if coreFont == CoreFont.HELVETICA_BOLD_OBLIQUE {
            self.name = Helvetica_BoldOblique.name
            self.bBoxLLx = Helvetica_BoldOblique.bBoxLLx
            self.bBoxLLy = Helvetica_BoldOblique.bBoxLLy
            self.bBoxURx = Helvetica_BoldOblique.bBoxURx
            self.bBoxURy = Helvetica_BoldOblique.bBoxURy
            self.underlinePosition = Helvetica_BoldOblique.underlinePosition
            self.underlineThickness = Helvetica_BoldOblique.underlineThickness
            self.metrics = Helvetica_BoldOblique.metrics
        } else if coreFont == CoreFont.TIMES_ROMAN {
            self.name = Times_Roman.name
            self.bBoxLLx = Times_Roman.bBoxLLx
            self.bBoxLLy = Times_Roman.bBoxLLy
            self.bBoxURx = Times_Roman.bBoxURx
            self.bBoxURy = Times_Roman.bBoxURy
            self.underlinePosition = Times_Roman.underlinePosition
            self.underlineThickness = Times_Roman.underlineThickness
            self.metrics = Times_Roman.metrics
        } else if coreFont == CoreFont.TIMES_BOLD {
            self.name = Times_Bold.name
            self.bBoxLLx = Times_Bold.bBoxLLx
            self.bBoxLLy = Times_Bold.bBoxLLy
            self.bBoxURx = Times_Bold.bBoxURx
            self.bBoxURy = Times_Bold.bBoxURy
            self.underlinePosition = Times_Bold.underlinePosition
            self.underlineThickness = Times_Bold.underlineThickness
            self.metrics = Times_Bold.metrics
        } else if coreFont == CoreFont.TIMES_ITALIC {
            self.name = Times_Italic.name
            self.bBoxLLx = Times_Italic.bBoxLLx
            self.bBoxLLy = Times_Italic.bBoxLLy
            self.bBoxURx = Times_Italic.bBoxURx
            self.bBoxURy = Times_Italic.bBoxURy
            self.underlinePosition = Times_Italic.underlinePosition
            self.underlineThickness = Times_Italic.underlineThickness
            self.metrics = Times_Italic.metrics
        } else if coreFont == CoreFont.TIMES_BOLD_ITALIC {
            self.name = Times_BoldItalic.name
            self.bBoxLLx = Times_BoldItalic.bBoxLLx
            self.bBoxLLy = Times_BoldItalic.bBoxLLy
            self.bBoxURx = Times_BoldItalic.bBoxURx
            self.bBoxURy = Times_BoldItalic.bBoxURy
            self.underlinePosition = Times_BoldItalic.underlinePosition
            self.underlineThickness = Times_BoldItalic.underlineThickness
            self.metrics = Times_BoldItalic.metrics
        } else if coreFont == CoreFont.SYMBOL {
            self.name = Symbol.name
            self.bBoxLLx = Symbol.bBoxLLx
            self.bBoxLLy = Symbol.bBoxLLy
            self.bBoxURx = Symbol.bBoxURx
            self.bBoxURy = Symbol.bBoxURy
            self.underlinePosition = Symbol.underlinePosition
            self.underlineThickness = Symbol.underlineThickness
            self.metrics = Symbol.metrics
        } else if coreFont == CoreFont.ZAPF_DINGBATS {
            self.name = ZapfDingbats.name
            self.bBoxLLx = ZapfDingbats.bBoxLLx
            self.bBoxLLy = ZapfDingbats.bBoxLLy
            self.bBoxURx = ZapfDingbats.bBoxURx
            self.bBoxURy = ZapfDingbats.bBoxURy
            self.underlinePosition = ZapfDingbats.underlinePosition
            self.underlineThickness = ZapfDingbats.underlineThickness
            self.metrics = ZapfDingbats.metrics
        }
    }
}

File: com/pdfjet/CoreFont.java
package com.pdfjet;

import com.pdfjet.corefonts.*;

/**
 * The metrics of the 14 standard PDF fonts.
 */
public class CoreFont {
    /** Courier. */
    public static final int COURIER = 1;
    /** Courier Bold. */
    public static final int COURIER_BOLD = 2;
    /** Courier Oblique. */
    public static final int COURIER_OBLIQUE = 3;
    /** Courier Bold Oblique. */
    public static final int COURIER_BOLD_OBLIQUE = 4;
    /** Helvetica. */
    public static final int HELVETICA = 5;
    /** Helvetica Bold. */
    public static final int HELVETICA_BOLD = 6;
    /** Helvetica Oblique. */
    public static final int HELVETICA_OBLIQUE = 7;
    /** Helvetica Bold Oblique. */
    public static final int HELVETICA_BOLD_OBLIQUE = 8;
    /** Times Roman. */
    public static final int TIMES_ROMAN = 9;
    /** Times Bold. */
    public static final int TIMES_BOLD = 10;
    /** Times Italic. */
    public static final int TIMES_ITALIC = 11;
    /** Times Bold Italic. */
    public static final int TIMES_BOLD_ITALIC = 12;
    /** Symbol. */
    public static final int SYMBOL = 13;
    /** Zapf Dingbats. */
    public static final int ZAPF_DINGBATS = 14;

    /** The PostScript name of the font. */
    protected String name;
    /** The x coordinate of the lower left corner of the font bounding box. */
    protected int bBoxLLx;
    /** The y coordinate of the lower left corner of the font bounding box. */
    protected int bBoxLLy;
    /** The x coordinate of the upper right corner of the font bounding box. */
    protected int bBoxURx;
    /** The y coordinate of the upper right corner of the font bounding box. */
    protected int bBoxURy;
    /** The position of the underline relative to the baseline. */
    protected int underlinePosition;
    /** The thickness of the underline. */
    protected int underlineThickness;
    /** The character widths and kerning pairs. */
    protected int[][] metrics;

    /**
     * Loads the metrics of the specified standard font.
     *
     * @param coreFont one of the font constants in this class, for example CoreFont.HELVETICA.
     */
    public CoreFont(int coreFont) {
        switch (coreFont) {
            case 1: // CoreFont.COURIER
            this.name = Courier.name;
            this.bBoxLLx = Courier.bBoxLLx;
            this.bBoxLLy = Courier.bBoxLLy;
            this.bBoxURx = Courier.bBoxURx;
            this.bBoxURy = Courier.bBoxURy;
            this.underlinePosition = Courier.underlinePosition;
            this.underlineThickness = Courier.underlineThickness;
            this.metrics = Courier.metrics;
            break;

            case 2: // CoreFont.COURIER_BOLD
            this.name = Courier_Bold.name;
            this.bBoxLLx = Courier_Bold.bBoxLLx;
            this.bBoxLLy = Courier_Bold.bBoxLLy;
            this.bBoxURx = Courier_Bold.bBoxURx;
            this.bBoxURy = Courier_Bold.bBoxURy;
            this.underlinePosition = Courier_Bold.underlinePosition;
            this.underlineThickness = Courier_Bold.underlineThickness;
            this.metrics = Courier_Bold.metrics;
            break;

            case 3: // CoreFont.COURIER_OBLIQUE
            this.name = Courier_Oblique.name;
            this.bBoxLLx = Courier_Oblique.bBoxLLx;
            this.bBoxLLy = Courier_Oblique.bBoxLLy;
            this.bBoxURx = Courier_Oblique.bBoxURx;
            this.bBoxURy = Courier_Oblique.bBoxURy;
            this.underlinePosition = Courier_Oblique.underlinePosition;
            this.underlineThickness = Courier_Oblique.underlineThickness;
            this.metrics = Courier_Oblique.metrics;
            break;

            case 4: // CoreFont.COURIER_BOLD_OBLIQUE
            this.name = Courier_BoldOblique.name;
            this.bBoxLLx = Courier_BoldOblique.bBoxLLx;
            this.bBoxLLy = Courier_BoldOblique.bBoxLLy;
            this.bBoxURx = Courier_BoldOblique.bBoxURx;
            this.bBoxURy = Courier_BoldOblique.bBoxURy;
            this.underlinePosition = Courier_BoldOblique.underlinePosition;
            this.underlineThickness = Courier_BoldOblique.underlineThickness;
            this.metrics = Courier_BoldOblique.metrics;
            break;

            case 5: // case CoreFont.HELVETICA:
            this.name = Helvetica.name;
            this.bBoxLLx = Helvetica.bBoxLLx;
            this.bBoxLLy = Helvetica.bBoxLLy;
            this.bBoxURx = Helvetica.bBoxURx;
            this.bBoxURy = Helvetica.bBoxURy;
            this.underlinePosition = Helvetica.underlinePosition;
            this.underlineThickness = Helvetica.underlineThickness;
            this.metrics = Helvetica.metrics;
            break;

            case 6: // CoreFont.HELVETICA_BOLD:
            this.name = Helvetica_Bold.name;
            this.bBoxLLx = Helvetica_Bold.bBoxLLx;
            this.bBoxLLy = Helvetica_Bold.bBoxLLy;
            this.bBoxURx = Helvetica_Bold.bBoxURx;
            this.bBoxURy = Helvetica_Bold.bBoxURy;
            this.underlinePosition = Helvetica_Bold.underlinePosition;
            this.underlineThickness = Helvetica_Bold.underlineThickness;
            this.metrics = Helvetica_Bold.metrics;
            break;

            case 7: // CoreFont.HELVETICA_OBLIQUE:
            this.name = Helvetica_Oblique.name;
            this.bBoxLLx = Helvetica_Oblique.bBoxLLx;
            this.bBoxLLy = Helvetica_Oblique.bBoxLLy;
            this.bBoxURx = Helvetica_Oblique.bBoxURx;
            this.bBoxURy = Helvetica_Oblique.bBoxURy;
            this.underlinePosition = Helvetica_Oblique.underlinePosition;
            this.underlineThickness = Helvetica_Oblique.underlineThickness;
            this.metrics = Helvetica_Oblique.metrics;
            break;

            case 8: // CoreFont.HELVETICA_BOLD_OBLIQUE:
            this.name = Helvetica_BoldOblique.name;
            this.bBoxLLx = Helvetica_BoldOblique.bBoxLLx;
            this.bBoxLLy = Helvetica_BoldOblique.bBoxLLy;
            this.bBoxURx = Helvetica_BoldOblique.bBoxURx;
            this.bBoxURy = Helvetica_BoldOblique.bBoxURy;
            this.underlinePosition = Helvetica_BoldOblique.underlinePosition;
            this.underlineThickness = Helvetica_BoldOblique.underlineThickness;
            this.metrics = Helvetica_BoldOblique.metrics;
            break;

            case 9: // CoreFont.TIMES_ROMAN:
            this.name = Times_Roman.name;
            this.bBoxLLx = Times_Roman.bBoxLLx;
            this.bBoxLLy = Times_Roman.bBoxLLy;
            this.bBoxURx = Times_Roman.bBoxURx;
            this.bBoxURy = Times_Roman.bBoxURy;
            this.underlinePosition = Times_Roman.underlinePosition;
            this.underlineThickness = Times_Roman.underlineThickness;
            this.metrics = Times_Roman.metrics;
            break;

            case 10: // CoreFont.TIMES_BOLD:
            this.name = Times_Bold.name;
            this.bBoxLLx = Times_Bold.bBoxLLx;
            this.bBoxLLy = Times_Bold.bBoxLLy;
            this.bBoxURx = Times_Bold.bBoxURx;
            this.bBoxURy = Times_Bold.bBoxURy;
            this.underlinePosition = Times_Bold.underlinePosition;
            this.underlineThickness = Times_Bold.underlineThickness;
            this.metrics = Times_Bold.metrics;
            break;

            case 11: // CoreFont.TIMES_ITALIC:
            this.name = Times_Italic.name;
            this.bBoxLLx = Times_Italic.bBoxLLx;
            this.bBoxLLy = Times_Italic.bBoxLLy;
            this.bBoxURx = Times_Italic.bBoxURx;
            this.bBoxURy = Times_Italic.bBoxURy;
            this.underlinePosition = Times_Italic.underlinePosition;
            this.underlineThickness = Times_Italic.underlineThickness;
            this.metrics = Times_Italic.metrics;
            break;

            case 12: // CoreFont.TIMES_BOLD_ITALIC:
            this.name = Times_BoldItalic.name;
            this.bBoxLLx = Times_BoldItalic.bBoxLLx;
            this.bBoxLLy = Times_BoldItalic.bBoxLLy;
            this.bBoxURx = Times_BoldItalic.bBoxURx;
            this.bBoxURy = Times_BoldItalic.bBoxURy;
            this.underlinePosition = Times_BoldItalic.underlinePosition;
            this.underlineThickness = Times_BoldItalic.underlineThickness;
            this.metrics = Times_BoldItalic.metrics;
            break;

            case 13: // CoreFont.SYMBOL:
            this.name = Symbol.name;
            this.bBoxLLx = Symbol.bBoxLLx;
            this.bBoxLLy = Symbol.bBoxLLy;
            this.bBoxURx = Symbol.bBoxURx;
            this.bBoxURy = Symbol.bBoxURy;
            this.underlinePosition = Symbol.underlinePosition;
            this.underlineThickness = Symbol.underlineThickness;
            this.metrics = Symbol.metrics;
            break;

            case 14: // CoreFont.ZAPF_DINGBATS:
            this.name = ZapfDingbats.name;
            this.bBoxLLx = ZapfDingbats.bBoxLLx;
            this.bBoxLLy = ZapfDingbats.bBoxLLy;
            this.bBoxURx = ZapfDingbats.bBoxURx;
            this.bBoxURy = ZapfDingbats.bBoxURy;
            this.underlinePosition = ZapfDingbats.underlinePosition;
            this.underlineThickness = ZapfDingbats.underlineThickness;
            this.metrics = ZapfDingbats.metrics;
            break;
        }
    }
}
